opencog/atoms/proto/ValueFactory.h

```diff
@@ -0,0 +1,98 @@
+#ifndef _VALUE_FACTORY_H_
+#define _VALUE_FACTORY_H_
+
+#include "ProtoAtom.h"
+#include <opencog/atoms/base/Handle.h>
+
+#include <exception>
+#include <map>
+#include <typeinfo>
+#include <typeindex>
+#include <vector>
+
+namespace opencog
+{
+using CreateProto = ProtoAtomPtr (*) (...);
+
+struct FuncRegister
+{
+    CreateProto func;
+    std::vector<std::type_index> args;
+};
+
+class ValueFactory
+{
+private:
+
+#define THROW_ERROR \
+    (throw std::invalid_argument("There was no function found for the given type."))
+
+    ValueFactory() {}
+
+    std::map<Type, std::vector<FuncRegister>> func_register;
+    std::map<Type, CreateProto> cast_register;
+
+public:
+
+    friend ValueFactory& valuefactory();
+
+    /** Registers the creator functions.
+     *  @param vtype The value type.
+     *  @param func a pointer to the creator function of the value.
+     *  @param args  an ordered vector the types of arguments the creator takes.
+     */
+     void register_factory(Type vtype, CreateProto func, std::vector<std::type_index> args);
```

Comment:
Change name to be consistent with `ClassServer.h`, where this is called `addFactory()`